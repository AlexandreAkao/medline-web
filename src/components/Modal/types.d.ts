interface IModalStyles {
  customStyles?: import('styled-components').Interpolation<import('react').CSSProperties>;
}

interface IModalProps extends IModalStyles {
  children: React.ReactNode;
  isVisible: boolean;
  onClose?: () => void;
  onOpen?: () => void;
}
