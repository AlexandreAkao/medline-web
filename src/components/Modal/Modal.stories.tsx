import { action } from '@storybook/addon-actions';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Modal from 'components/Modal';
import colors from 'styles/colors';

export default {
  title: 'Modal',
  args: {},
  component: Modal,
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary.normal,
          width: '100%',
          height: '100%',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = args => <Modal {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  isVisible: true,
  customStyles: {},
  onOpen: action('Open'),
  onClose: action('Close'),
  children: <div>Any children</div>,
};
