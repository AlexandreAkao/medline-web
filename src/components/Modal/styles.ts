import styled, { keyframes } from 'styled-components';

import colors from 'styles/colors';
import zIndexPriority from 'styles/zIndexPriority';
import { getColorAlpha } from 'utils/colors';

const showModal = keyframes`
  0% {
    transform: scale(0);
  }
  80% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
`;

export const ModalContainer = styled.div`
  z-index: ${zIndexPriority.ALMOST_MAX};
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: ${getColorAlpha(colors.black.soft, 0.6)};
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const ModalContent = styled.div<IModalStyles>`
  background-color: ${colors.white.normal};
  border-radius: 0.75rem;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.3s ease-in-out;
  animation: ${showModal} 0.3s ease-in-out;

  ${({ customStyles }) => customStyles}
`;
