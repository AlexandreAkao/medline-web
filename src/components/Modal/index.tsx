import { useEffect, useRef } from 'react';

import { ModalContainer, ModalContent } from 'components/Modal/styles';
import Portal from 'components/Portal';
import useOutsideHandler from 'hooks/useOutsideHandler';

function Modal({ isVisible, onClose, onOpen, customStyles, children }: IModalProps) {
  const ref = useRef(null);
  const refContent = useRef(null);

  useOutsideHandler<HTMLDivElement>(refContent, onClose);

  useEffect(() => {
    const closeOnEscapeKey = (e: KeyboardEvent) => (e.key === 'Escape' && onClose && isVisible ? onClose() : null);
    document.body.addEventListener('keydown', closeOnEscapeKey);

    return () => {
      document.body.removeEventListener('keydown', closeOnEscapeKey);
    };
  }, [isVisible, onClose]);

  useEffect(() => {
    if (isVisible && onOpen) onOpen();
  }, [isVisible, onOpen]);

  if (!isVisible) return null;

  return (
    <Portal wrapperId="portal-modal">
      <ModalContainer ref={ref}>
        <ModalContent ref={refContent} customStyles={customStyles}>
          {children}
        </ModalContent>
      </ModalContainer>
    </Portal>
  );
}

export default Modal;
