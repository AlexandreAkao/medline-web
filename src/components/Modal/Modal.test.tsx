import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import { vi } from 'vitest';
import userEvent from '@testing-library/user-event';

import * as stories from 'components/Modal/Modal.stories';
import Modal from 'components/Modal';

const { Primary } = composeStories(stories);

describe('Modal', () => {
  const renderComponent = (customProps: Partial<IModalProps> = {}) => {
    const props: IModalProps = {
      isVisible: true,
      children: 'modal example',
      ...customProps,
    };
    return render(<Modal {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a Modal', () => {
    render(<Primary />);
    const modalElement = screen.getByTestId(/test-modal-container/i);
    expect(modalElement).toBeInTheDocument();
  });

  it('should not render Modal', () => {
    const { container } = renderComponent({ isVisible: false });
    expect(container.childElementCount).toBeFalsy();
  });

  it('should call onOpen when open Modal', () => {
    const onOpen = vi.fn();
    renderComponent({ onOpen });
    expect(onOpen).toBeCalled();
  });

  it('should call onClose when close Modal', async () => {
    const onClose = vi.fn();
    renderComponent({ onClose });
    const modalElement = screen.getByTestId(/test-modal-container/i);

    await userEvent.click(modalElement);

    expect(onClose).toBeCalled();
  });
});
