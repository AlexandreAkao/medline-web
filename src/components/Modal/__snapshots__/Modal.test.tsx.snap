// Vitest Snapshot v1

exports[`Modal > should match with snapshot 1`] = `<div />`;
